crates/sui-types/src/sui_system_state.rs

```diff
@@ -553,6 +581,104 @@ impl Default for SuiSystemState {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq, JsonSchema)]
+pub enum SuiSystemState {
+    V1(SuiSystemStateInnerV1),
+}
+
+impl SuiSystemState {
+    /// Always return the initial (genesis) version of the system state object.
+    /// Genesis always uses this version regardless of the current version.
+    pub fn into_genesis_version(self) -> SuiSystemStateInnerV1 {
+        match self {
+            SuiSystemState::V1(inner) => inner,
+        }
+    }
+
+    /// Always treat this as the latest version and return it. If it's not, return Err.
+    /// NOTE: Update the return type of this function whenever we upgrade the system state type.
+    pub fn to_latest_version(self) -> SuiResult<SuiSystemStateInnerV1> {
+        match self {
+            SuiSystemState::V1(inner) => Ok(inner),
+            // If this is not the latest version, we return None.
+            // Uncomment once we have more than 1 versions.
+            // _ => Err(SuiError::SuiSystemStateUnexpectedVersion),
+        }
+    }
+
+    pub fn epoch(&self) -> u64 {
+        match self {
+            SuiSystemState::V1(inner) => inner.epoch,
+        }
+    }
+
+    pub fn reference_gas_price(&self) -> u64 {
+        match self {
+            SuiSystemState::V1(inner) => inner.reference_gas_price,
+        }
+    }
+
+    pub fn epoch_start_timestamp_ms(&self) -> u64 {
+        match self {
+            SuiSystemState::V1(inner) => inner.epoch_start_timestamp_ms,
+        }
+    }
+
+    pub fn safe_mode(&self) -> bool {
+        match self {
+            SuiSystemState::V1(inner) => inner.safe_mode,
+        }
+    }
+
+    pub fn get_current_epoch_committee(&self) -> CommitteeWithNetAddresses {
+        match self {
+            SuiSystemState::V1(inner) => inner.get_current_epoch_committee(),
+        }
+    }
+
+    pub fn get_current_epoch_narwhal_committee(&self) -> NarwhalCommittee {
+        match self {
+            SuiSystemState::V1(inner) => inner.get_current_epoch_narwhal_committee(),
+        }
+    }
+
+    pub fn get_current_epoch_narwhal_worker_cache(
+        &self,
+        transactions_address: &Multiaddr,
+    ) -> WorkerCache {
+        match self {
+            SuiSystemState::V1(inner) => {
+                inner.get_current_epoch_narwhal_worker_cache(transactions_address)
+            }
+        }
+    }
+
+    pub fn get_validator_metadata_vec(&self) -> Vec<ValidatorMetadata> {
+        match self {
+            SuiSystemState::V1(inner) => inner.get_validator_metadata_vec(),
+        }
+    }
+
+    pub fn get_current_epoch_authority_names_to_peer_ids(&self) -> HashMap<AuthorityName, PeerId> {
+        match self {
+            SuiSystemState::V1(inner) => inner.get_current_epoch_authority_names_to_peer_ids(),
+        }
+    }
+
+    pub fn new_for_testing(epoch: EpochId) -> Self {
+        SuiSystemState::V1(SuiSystemStateInnerV1 {
+            epoch,
+            ..Default::default()
+        })
+    }
```

Comment:
for all this boilerplate I might recommend using `enum_dispatch`. See https://github.com/MystenLabs/sui/pull/8754 for an example.